File: inspect_docx.py
# -*- coding: utf-8 -*-
import zipfile
import xml.etree.ElementTree as ET
import sys
import os
import traceback

_base = os.path.dirname(os.path.abspath(__file__))
os.chdir(_base)  # 결과 파일이 항상 스크립트 폴더에 생성되도록
_cwd = os.getcwd()
_path_in_project = os.path.join(_base, "오이코스논문 신탬플릿---수정본.docx")
path = _path_in_project if os.path.exists(_path_in_project) else r"e:\##OIKOS(AI융합)-박사과정\##논문\##오이코스 공식 논문형식\오이코스논문 신탬플릿---수정본.docx"
if not os.path.exists(path):
    with open(os.path.join(_base, "error_log.txt"), "w", encoding="utf-8") as f:
        f.write(f"DOCX not found.\n_base={_base}\ncwd={_cwd}\npath={path}\n")
    print("Error: DOCX file not found. See error_log.txt")
    sys.exit(1)
ns = {"w": "http://schemas.openxmlformats.org/wordprocessingml/2006/main"}

def get_text(elem):
    parts = []
    for t in elem.iter("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}t"):
        if t.text:
            parts.append(t.text)
        if t.tail:
            parts.append(t.tail)
    return "".join(parts).strip()

def get_style_id(p):
    pPr = p.find("w:pPr", ns)
    if pPr is not None:
        pStyle = pPr.find("w:pStyle", ns)
        if pStyle is not None:
            return pStyle.get("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val")
    return None

def get_run_props(r):
    rPr = r.find("w:rPr", ns)
    if rPr is None:
        return {}
    out = {}
    b = rPr.find("w:b", ns)
    if b is not None: out["bold"] = True
    sz = rPr.find("w:sz", ns)
    if sz is not None: out["fontSize"] = sz.get("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val")
    return out

try:
    with zipfile.ZipFile(path, "r") as z:
        # 1) 스타일 정의 (styles.xml)
        style_names = {}
        if "word/styles.xml" in z.namelist():
            with z.open("word/styles.xml") as f:
                stree = ET.parse(f)
                sroot = stree.getroot()
                for style in sroot.findall(".//{http://schemas.openxmlformats.org/wordprocessingml/2006/main}style"):
                    sid = style.get("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}styleId")
                    name = style.find("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}name")
                    if sid and name is not None and name.get("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val"):
                        style_names[sid] = name.get("{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val")

        # 2) 본문 문단 + 스타일
        with z.open("word/document.xml") as f:
            tree = ET.parse(f)
            root = tree.getroot()
            body = root.find("w:body", ns)
            if body is None:
                body = root

            out_lines = []
            for p in body.findall(".//w:p", ns):
                line = get_text(p)
                if not line:
                    continue
                style_id = get_style_id(p)
                style_name = style_names.get(style_id, style_id or "(없음)")
                out_lines.append((style_name, line))

    # 결과를 파일로 저장 (스크립트 폴더 + 현재 디렉터리)
    out_path = os.path.join(_base, "template_structure.txt")
    with open(out_path, "w", encoding="utf-8") as out:
        out.write("=== 스타일 정의 (styles.xml) ===\n")
        for sid, name in sorted(style_names.items()):
            out.write(f"  {sid} -> {name}\n")
        out.write("\n=== 문단 목록 (스타일 | 내용) ===\n")
        for style_name, line in out_lines:
            if len(line) > 70:
                line = line[:70] + "..."
            out.write(f"  [{style_name}] {line}\n")
    # Cursor 프로젝트에서 찾을 수 있도록 현재 디렉터리에도 복사
    if _cwd != _base:
        try:
            import shutil
            shutil.copy(out_path, os.path.join(_cwd, "template_structure.txt"))
        except Exception:
            pass
    print("Done. See template_structure.txt")
except Exception as e:
    err_path = os.path.join(_base, "error_log.txt")
    with open(err_path, "w", encoding="utf-8") as f:
        f.write(traceback.format_exc())
    print("Error. See error_log.txt")
    raise

File: inspect_docx.js
// Node로 docx 구조 분석 (Python 불필요)
const fs = require("fs");
const path = require("path");

const base = __dirname;
const docxPath = path.join(base, "오이코스논문 신탬플릿---수정본.docx");
const outPath = path.join(base, "template_structure.txt");

if (!fs.existsSync(docxPath)) {
  fs.writeFileSync(path.join(base, "error_log.txt"), "DOCX file not found: " + docxPath, "utf8");
  console.log("Error: DOCX not found. See error_log.txt");
  process.exit(1);
}

// docx = zip. Node에 zip 내장 없음 → adm-zip 사용 (npm i adm-zip)
let AdmZip;
try {
  AdmZip = require("adm-zip");
} catch (e) {
  console.log("먼저 실행: npm install adm-zip");
  process.exit(1);
}

const zip = new AdmZip(docxPath);
const zipEntries = zip.getEntries();

const NS = "http://schemas.openxmlformats.org/wordprocessingml/2006/main";
function getText(el) {
  const list = el.getElementsByTagNameNS(NS, "t");
  let s = "";
  for (let i = 0; i < list.length; i++) {
    if (list[i].textContent) s += list[i].textContent;
  }
  return s.trim();
}
function getStyleId(p) {
  const pPr = p.getElementsByTagNameNS(NS, "pPr")[0];
  if (!pPr) return null;
  const pStyle = pPr.getElementsByTagNameNS(NS, "pStyle")[0];
  if (!pStyle) return null;
  return pStyle.getAttribute("w:val") || pStyle.getAttribute("val");
}

const styleNames = {};
const styleEntry = zipEntries.find((e) => e.entryName === "word/styles.xml");
if (styleEntry) {
  const xml = styleEntry.getData().toString("utf8");
  const styleDoc = new (require("xmldom").DOMParser)().parseFromString(xml);
  const styles = styleDoc.getElementsByTagNameNS(NS, "style");
  for (let i = 0; i < styles.length; i++) {
    const sid = styles[i].getAttribute("w:styleId") || styles[i].getAttribute("styleId");
    const nameEl = styles[i].getElementsByTagNameNS(NS, "name")[0];
    if (sid && nameEl) {
      const name = nameEl.getAttribute("w:val") || nameEl.getAttribute("val");
      if (name) styleNames[sid] = name;
    }
  }
}

const docEntry = zipEntries.find((e) => e.entryName === "word/document.xml");
if (!docEntry) {
  fs.writeFileSync(outPath, "word/document.xml not found in DOCX\n", "utf8");
  console.log("Done. See template_structure.txt");
  process.exit(0);
}

const docXml = docEntry.getData().toString("utf8");
const docDoc = new (require("xmldom").DOMParser)().parseFromString(docXml);
const body = docDoc.getElementsByTagNameNS(NS, "body")[0] || docDoc.documentElement;
const paragraphs = body.getElementsByTagNameNS(NS, "p");

const outLines = [];
for (let i = 0; i < paragraphs.length; i++) {
  const line = getText(paragraphs[i]);
  if (!line) continue;
  const styleId = getStyleId(paragraphs[i]);
  const styleName = styleNames[styleId] || styleId || "(없음)";
  outLines.push([styleName, line]);
}

let out = "=== 스타일 정의 (styles.xml) ===\n";
Object.keys(styleNames)
  .sort()
  .forEach((sid) => {
    out += "  " + sid + " -> " + styleNames[sid] + "\n";
  });
out += "\n=== 문단 목록 (스타일 | 내용) ===\n";
outLines.forEach(([styleName, line]) => {
  if (line.length > 70) line = line.slice(0, 70) + "...";
  out += "  [" + styleName + "] " + line + "\n";
});

fs.writeFileSync(outPath, out, "utf8");
console.log("Done. See template_structure.txt");
